### PR TITLE
Lift trivial transpiler passes to handle control flow

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

This converts all transpiler passes whose structure allows their `run` methods to be made trivially recursive, either by eagerly modifying all control-flow operations in a depth-first pass before running the method on the top level of the circuit, or by inserting the recursion during an interior loop.

There are other passes that are relatively simple to handle as well (`UnitarySynthesis` is one), but their current structure would require just a shade more work than the trivial cases in this commit.

### Details and comments

Close #8445 (by obsoleting it).  We can rework #8565 to do essentially this as well - we probably just want to take a bit more care than that PR currently does not to redo all the setup work in the pass.

### CODE REVIEW

File: test/python/transpiler/test_linear_functions_passes.py [hunk 1]
```diff
@@ -329,6 +329,53 @@ def test_connected_gates(self):
...
+
+        test = QuantumCircuit(4, 1)
+        test.h(0)
+        test.measure(0, 0)
+        test.if_else((0, True), circuit.copy(), circuit.copy(), range(4), [0])
+
+        expected = QuantumCircuit(4, 1)
+        expected.h(0)
+        expected.measure(0, 0)
+        expected.if_else((0, True), pass_(circuit), pass_(circuit), range(4), [0])
```
Comment (ewinston) [inline test/python/transpiler/test_linear_functions_passes.py:349]: It seems like this pattern of creating the expected control flow blocks by calling the pass on them when creating the control flow instruction could be repeated in some of the other tests e.g. `test_optimize_swap_before_measure`, `test_cx_cancellation`, `test_clifford_passes`, etc.

**Review by ewinston — COMMENTED** [review]
Overall this LGTM. Just a minor comment about tests.

File: test/python/transpiler/test_linear_functions_passes.py (hunk 1 above)
Comment (jakelishman) [reply test/python/transpiler/test_linear_functions_passes.py:349]: I actually swapped from this style to the pure style about halfway through - I thought being _completely_ explicit about the expected form might be neater. I'm happy to go either way - which do you prefer?

Comment (ewinston) [reply test/python/transpiler/test_linear_functions_passes.py:349]: I used explicit in my tests but I thought this was a nice approach. One has to be a little careful with nested statements but perhaps no more than being explicit.

**Review by ewinston — APPROVED** [review]

File: test/python/transpiler/test_linear_functions_passes.py (hunk 1 above)
Comment (ewinston) [reply test/python/transpiler/test_linear_functions_passes.py:349]: I approved since we can perhaps worry about it later if it comes up.

**Pull request merged**